Index: b/Makefile.in
===================================================================
--- a/Makefile.in
+++ b/Makefile.in
@@ -380,13 +380,13 @@ install_doc::
 
 
 build_docbook::
-	cd doc/docbook; make
+	$(MAKE) -C doc/docbook
 
 dl_docbook::
-	cd doc/docbook; make dl_docs
+	$(MAKE) -C doc/docbook dl_docs
 
 install_docbook: build_docbook
-	cd doc/docbook; make install
+	$(MAKE) -C doc/docbook install
 
 install_man::
 	-mkdir -p $(DESTDIR)$(man1dir)
