#DPATCHLEVEL=1

diff -Naur bochs-2.2.orig/gdbstub.cc bochs-2.2/gdbstub.cc
--- bochs-2.2.orig/gdbstub.cc	2005-03-16 18:36:31.000000000 +0200
+++ bochs-2.2/gdbstub.cc	2005-07-07 19:51:04.000000000 +0300
@@ -368,7 +368,7 @@
        valid = access_linear(laddress,
                              len + (laddress & 0xfff) - 4096,
                              rw,
-                             (Bit8u *)((unsigned int)data + 
+                             (Bit8u *)((unsigned long)data + 
                                       4096 - (laddress & 0xfff)));
        return(valid);
      }
diff -Naur bochs-2.2.orig/gui/wx.cc bochs-2.2/gui/wx.cc
--- bochs-2.2.orig/gui/wx.cc	2004-11-06 19:03:43.000000000 +0200
+++ bochs-2.2/gui/wx.cc	2005-07-07 19:51:04.000000000 +0300
@@ -101,7 +101,7 @@
 static Bit8u h_panning = 0, v_panning = 0;
 static Bit16u line_compare = 1023;
 static unsigned vga_bpp=8;
-struct {
+static struct {
   unsigned char red;
   unsigned char green;
   unsigned char blue;
diff -Naur bochs-2.2.orig/gui/x.cc bochs-2.2/gui/x.cc
--- bochs-2.2.orig/gui/x.cc	2005-04-29 22:06:24.000000000 +0300
+++ bochs-2.2/gui/x.cc	2005-07-07 19:51:04.000000000 +0300
@@ -128,7 +128,7 @@
 
 static Pixmap vgafont[256];
 
-struct {
+static struct {
   Pixmap bmap;
   unsigned xdim;
   unsigned ydim;
diff -Naur bochs-2.2.orig/iodev/eth_linux.cc bochs-2.2/iodev/eth_linux.cc
--- bochs-2.2.orig/iodev/eth_linux.cc	2004-10-07 20:38:03.000000000 +0300
+++ bochs-2.2/iodev/eth_linux.cc	2005-07-07 19:51:04.000000000 +0300
@@ -215,7 +215,7 @@
   this->filter[3].k = (macaddr[0] & 0xff) << 8 | (macaddr[1] & 0xff);
   fp.len = BX_LSF_ICNT;
   fp.filter = this->filter;
-  BX_INFO(("eth_linux: fp.len=%d fp.filter=%x", fp.len, (unsigned) fp.filter));
+  BX_INFO(("eth_linux: fp.len=%d fp.filter=%lx", fp.len, (unsigned long) fp.filter));
   if (setsockopt(this->fd, SOL_SOCKET, SO_ATTACH_FILTER, &fp, sizeof(fp)) < 0) {
     BX_PANIC(("eth_linux: could not set socket filter: %s", strerror(errno)));
     close(this->fd);
