#DPATCHLEVEL=0
--- iodev/hdimage.h	2006-03-15 15:08:02.000000000 +0000
+++ iodev/hdimage.h	2006-03-15 15:10:18.000000000 +0000
@@ -272,16 +272,8 @@
  off_t total_size;
 
  void panic(const char * message);
- off_t
-#ifndef PARANOID
-       sparse_image_t::
-#endif
-                       get_physical_offset();
- void
-#ifndef PARANOID
-       sparse_image_t::
-#endif
-                       set_virtual_page(Bit32u new_virtual_page);
+ off_t get_physical_offset();
+ void set_virtual_page(Bit32u new_virtual_page);
  void read_header();
  ssize_t read_page_fragment(Bit32u read_virtual_page, Bit32u read_page_offset, size_t read_size, void * buf);
 

