Index: b/bios/Makefile.in
===================================================================
--- a/bios/Makefile.in	2008-01-23 09:45:31.000000000 +0200
+++ b/bios/Makefile.in	2008-01-23 09:48:36.000000000 +0200
@@ -58,7 +58,7 @@ BIOS_BUILD_DATE = "-DBIOS_BUILD_DATE=\"$
 	$(CXX) -c $(BX_INCDIRS) $(CXXFLAGS) $(LOCAL_CXXFLAGS) @CXXFP@$< @OFP@$@
 
 
-bios: biossums BIOS-bochs-latest BIOS-bochs-legacy 
+bios: biossums BIOS-bochs-latest BIOS-bochs-legacy BIOS-qemu-latest
 
 clean:
 	@RMCOMMAND@ *.o *.a *.s _rombios*_.c rombios*.txt rombios*.sym
@@ -70,6 +70,7 @@ dist-clean: clean
 
 bios-clean:
 	@RMCOMMAND@ BIOS-bochs-*
+	@RMCOMMAND@ BIOS-qemu-*
 
 BIOS-bochs-legacy: rombios.c apmbios.S biossums rombios.h
 	$(GCC32) $(BIOS_BUILD_DATE) -DLEGACY -E -P $< > _rombiosl_.c
@@ -81,6 +82,15 @@ BIOS-bochs-legacy: rombios.c apmbios.S b
 	./biossums $@
 	@RMCOMMAND@ _rombiosl_.s
 
+rombios16-qemu.bin: rombios.c apmbios.S biossums rombios.h
+	$(GCC) $(BIOS_BUILD_DATE) -DBX_QEMU -E -P $< > _rombiosq_.c
+	$(BCC) -o rombiosq.s -C-c -D__i86__ -0 -S _rombiosq_.c
+	sed -e 's/^\.text//' -e 's/^\.data//' rombiosq.s > _rombiosq_.s
+	$(AS86) _rombiosq_.s -b $@.tmp.bin -u- -w- -g -0 -j -O -l rombiosq.txt
+	-perl ${srcdir}/makesym.perl < rombiosq.txt > rombiosq.sym
+	mv $@.tmp.bin $@
+	./biossums $@
+	@RMCOMMAND@ _rombiosq_.s
 
 rombios16.bin: rombios.c apmbios.S biossums rombios.h
 	$(GCC32) $(BIOS_BUILD_DATE) -E -P $< > _rombios_.c
@@ -97,12 +107,22 @@ rombios32.bin: rombios32.out rombios.h
 	objcopy -O binary $< $@
 	./biossums -pad $@
 
+rombios32-qemu.bin: rombios32-qemu.out rombios.h
+	objcopy -O binary $< $@
+	./biossums -pad $@
+
 rombios32.out: rombios32start.o rombios32.o rombios32.ld
 	ld -o $@ -T rombios32.ld rombios32start.o rombios32.o
 
+rombios32-qemu.out: rombios32start.o rombios32-qemu.o rombios32.ld
+	ld -o $@ -T rombios32.ld rombios32start.o rombios32-qemu.o
+
 rombios32.o: rombios32.c acpi-dsdt.hex
 	$(GCC32) -O2 -Wall -c -o $@ $<
 
+rombios32-qemu.o: rombios32.c acpi-dsdt.hex
+	$(GCC32) -DBX_QEMU -O2 -Wall -c -o $@ $<
+
 ifeq ("1", "0")
 acpi-dsdt.hex: acpi-dsdt.dsl
 	iasl -tc -p $@ $<
@@ -114,5 +134,8 @@ rombios32start.o: rombios32start.S
 BIOS-bochs-latest: rombios16.bin rombios32.bin
 	cat rombios32.bin rombios16.bin > $@
 
+BIOS-qemu-latest: rombios16-qemu.bin rombios32-qemu.bin
+	cat rombios32-qemu.bin rombios16-qemu.bin > $@
+
 biossums: biossums.c
 	$(GCC) -o biossums biossums.c
Index: b/bios/rombios32.c
===================================================================
--- a/bios/rombios32.c	2007-12-09 17:37:27.000000000 +0200
+++ b/bios/rombios32.c	2008-01-23 09:45:57.000000000 +0200
@@ -873,6 +873,11 @@ static void mptable_init(void)
     int ioapic_id, i, len;
     int mp_config_table_size;
 
+#ifdef BX_QEMU
+    if (smp_cpus <= 1)
+        return;
+#endif
+
 #ifdef BX_USE_EBDA_TABLES
     mp_config_table = (uint8_t *)(ram_size - ACPI_DATA_SIZE - MPTABLE_MAX_SIZE);
 #else
