---
 bios/Makefile.in |    2 --
 1 file changed, 2 deletions(-)

--- a/bios/Makefile.in
+++ b/bios/Makefile.in
@@ -122,13 +122,11 @@ rombios32.o: rombios32.c acpi-dsdt.hex r
 rombios32-qemu.o: rombios32.c acpi-dsdt.hex rombios.h
 	$(GCC32) -DBX_QEMU -O2 -Wall -c -o $@ $<
 
-ifeq ("1", "0")
 acpi-dsdt.hex: acpi-dsdt.dsl
 	cpp -P $< $<.i
 	iasl -tc -p $@ $<.i
 	rm $<.i
 	sed -i -e's/^unsigned/const unsigned/' $@
-endif
 
 rombios32start.o: rombios32start.S
 	$(GCC32) -c -o $@ $<
