Index: b/bios/Makefile.in
===================================================================
--- a/bios/Makefile.in	2007-11-20 06:29:59.000000000 +0200
+++ b/bios/Makefile.in	2007-11-20 06:30:07.000000000 +0200
@@ -102,10 +102,8 @@ rombios32.out: rombios32start.o rombios3
 rombios32.o: rombios32.c acpi-dsdt.hex
 	$(GCC) -O2 -Wall -c -o $@ $<
 
-ifeq ("1", "0")
 acpi-dsdt.hex: acpi-dsdt.dsl
 	iasl -tc -p $@ $<
-endif
 
 rombios32start.o: rombios32start.S
 	$(GCC) -c -o $@ $<
