Index: b/bios/Makefile.in
===================================================================
--- a/bios/Makefile.in	2007-12-07 08:04:28.000000000 +0200
+++ b/bios/Makefile.in	2007-12-07 08:05:12.000000000 +0200
@@ -122,10 +122,8 @@ rombios32.o: rombios32.c acpi-dsdt.hex
 rombios32-qemu.o: rombios32.c acpi-dsdt.hex
 	$(GCC) -DBX_QEMU -O2 -Wall -c -o $@ $<
 
-ifeq ("1", "0")
 acpi-dsdt.hex: acpi-dsdt.dsl
 	iasl -tc -p $@ $<
-endif
 
 rombios32start.o: rombios32start.S
 	$(GCC) -c -o $@ $<
