Index: b/bios/Makefile.in
===================================================================
--- a/bios/Makefile.in	2008-01-23 09:49:20.000000000 +0200
+++ b/bios/Makefile.in	2008-01-23 09:49:23.000000000 +0200
@@ -123,10 +123,8 @@ rombios32.o: rombios32.c acpi-dsdt.hex
 rombios32-qemu.o: rombios32.c acpi-dsdt.hex
 	$(GCC32) -DBX_QEMU -O2 -Wall -c -o $@ $<
 
-ifeq ("1", "0")
 acpi-dsdt.hex: acpi-dsdt.dsl
 	iasl -tc -p $@ $<
-endif
 
 rombios32start.o: rombios32start.S
 	$(GCC32) -c -o $@ $<
