---
 doc/docbook/user/user.dbk |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

--- a/doc/docbook/user/user.dbk
+++ b/doc/docbook/user/user.dbk
@@ -3860,7 +3860,7 @@ mouse usage at runtime (see <link linken
   The default value is 'ps2'. The other choices are 'imps2' (wheel mouse
   on PS/2), 'serial', 'serial_wheel' and 'serial_msys' (one com port requires
   setting 'mode=mouse', see <link linkend="bochsopt-com">com option</link>).
-  To connect a mouse to an USB port, see the <link linkend="bochsopt-usb_uhci">usb_uhci option</link>
+  To connect a mouse to an USB port, see the <link linkend="bochsopt-usb-uhci">usb_uhci option</link>
   (requires PCI and USB support).
 </para>
 </section>
@@ -3918,7 +3918,7 @@ located in the <constant>bochs/host/linu
 </para>
 </section>
 
-<section id="bochsopt-usb_uhci"><title>usb_uhci</title>
+<section id="bochsopt-usb-uhci"><title>usb_uhci</title>
 <para>
 Examples:
 <screen>
@@ -3957,7 +3957,7 @@ PCI support must be enabled.
 </para></note>
 </section>
 
-<section id="bochsopt-usb_ohci"><title>usb_ohci</title>
+<section id="bochsopt-usb-ohci"><title>usb_ohci</title>
 <para>
 Example:
 <screen>
@@ -3965,7 +3965,7 @@ Example:
 </screen>
 This option controls the presence of the USB OHCI host controller with a
 2-port hub. The portX option accepts the same device types with the same
-syntax as the UHCI controller (see the <link linkend="bochsopt-usb_uhci">usb_uhci option</link>).
+syntax as the UHCI controller (see the <link linkend="bochsopt-usb-uhci">usb_uhci option</link>).
 The OHCI HC must be assigned to a PCI slot.
 </para>
 </section>
