---
 doc/docbook/user/user.dbk |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/doc/docbook/user/user.dbk
+++ b/doc/docbook/user/user.dbk
@@ -3914,7 +3914,7 @@ mouse usage at runtime (see <link linken
   The default value is 'ps2'. The other choices are 'imps2' (wheel mouse
   on PS/2), 'serial', 'serial_wheel' and 'serial_msys' (one com port requires
   setting 'mode=mouse', see <link linkend="bochsopt-com">com option</link>).
-  To connect a mouse to an USB port, see the <link linkend="bochsopt-usb_uhci">usb_uhci option</link>
+  To connect a mouse to an USB port, see the <link linkend="bochsopt-usb-uhci">usb_uhci option</link>
   (requires PCI and USB support).
 </para>
 </section>
@@ -3972,7 +3972,7 @@ located in the <constant>bochs/host/linu
 </para>
 </section>
 
-<section id="bochsopt-usb_uhci"><title>usb_uhci</title>
+<section id="bochsopt-usb-uhci"><title>usb_uhci</title>
 <para>
 Examples:
 <screen>
