---
 gdbstub.cc |    2 ++
 1 file changed, 2 insertions(+)

--- a/gdbstub.cc
+++ b/gdbstub.cc
@@ -691,6 +691,7 @@ static void debug_loop(void)
       }
 
       case 'g':
+      {
 #if BX_SUPPORT_X86_64 == 0
         WriteHostDWordToLittleEndian(registers + 0, EAX);
         WriteHostDWordToLittleEndian(registers + 1, ECX);
@@ -762,6 +763,7 @@ static void debug_loop(void)
 #endif
         put_reply(obuf);
         break;
+      }
 
       case '?':
         sprintf(obuf, "S%02x", SIGTRAP);
