From 7a27e8bd24227824d63eecf9d2ecd881e4cb99c4 Mon Sep 17 00:00:00 2001
From: Guillem Jover <guillem@debian.org>
Date: Thu, 24 Jan 2008 03:55:12 +0200
Subject: [PATCH] Try to use pkg-config before relaying on AC_PATH_XTRA

---
 configure.in |   21 +++++++++++++++------
 1 files changed, 15 insertions(+), 6 deletions(-)

diff --git a/configure.in b/configure.in
index 7ae852c..bd41a2e 100644
--- a/configure.in
+++ b/configure.in
@@ -120,8 +120,6 @@ dnl else it disables shared libraries somehow.  The values from this macro
 dnl are written to ltdlconf.h.
 AC_LIB_LTDL
 
-AC_PATH_XTRA
-
 AC_C_BIGENDIAN
 AC_C_INLINE
 AC_CHECK_SIZEOF(unsigned char)
@@ -2216,11 +2214,22 @@ display_libs=""
 
 XPM_LIB=''
 if test "$with_x11" = yes; then
-  display_libs="$display_libs X11"
-  if test "$no_x" = yes; then
-    echo ERROR: X windows gui was selected, but X windows libraries were not found.
-    exit 1
+  PKG_CHECK_MODULES(X, x11,
+                    [x_found_with_pkgconfig=yes],
+                    [x_found_with_pkgconfig=no])
+
+  if test "$x_found_with_pkgconfig" = "yes"
+  then
+    :
+  else
+    AC_PATH_XTRA
+    if test "$no_x" = yes; then
+      echo ERROR: X windows gui was selected, but X windows libraries were not found.
+      exit 1
+    fi
   fi
+
+  display_libs="$display_libs X11"
   AC_DEFINE(BX_WITH_X11, 1)
   SPECIFIC_GUI_OBJS="$SPECIFIC_GUI_OBJS \$(GUI_OBJS_X11)"
   GUI_CFLAGS="$GUI_CFLAGS \$(X_CFLAGS)"
-- 
1.5.3.8

