Author: Guillem Jover <guillem@debian.org>

---
 bios/Makefile.in |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

--- a/bios/Makefile.in
+++ b/bios/Makefile.in
@@ -62,6 +62,7 @@ bios: biossums BIOS-bochs-latest BIOS-bo
 
 clean:
 	@RMCOMMAND@ *.o *.a *.s _rombios*_.c rombios*.txt rombios*.sym
+	@RMCOMMAND@ acpi-dsdt.hex acpi-dsdt.aml
 	@RMCOMMAND@ usage biossums
 	@RMCOMMAND@ rombios16.bin rombios16-qemu.bin
 	@RMCOMMAND@ rombios32.bin rombios32.out
@@ -124,13 +125,11 @@ rombios32.o: rombios32.c acpi-dsdt.hex r
 rombios32-qemu.o: rombios32.c acpi-dsdt.hex rombios.h
 	$(GCC32) -DBX_QEMU -O2 -Wall -c -o $@ $<
 
-ifeq ("1", "0")
 acpi-dsdt.hex: acpi-dsdt.dsl
 	cpp -P $< $<.i
 	iasl -tc -p $@ $<.i
 	rm $<.i
 	sed -i -e's/^unsigned/const unsigned/' $@
-endif
 
 rombios32start.o: rombios32start.S
 	$(GCC32) -c -o $@ $<
