diff -Nur bochs-2.1.1.old/Makefile.in bochs-2.1.1/Makefile.in
--- bochs-2.1.1.old/Makefile.in	2004-02-11 23:28:02.000000000 +0100
+++ bochs-2.1.1/Makefile.in	2004-03-24 23:14:47.000000000 +0100
@@ -70,7 +70,8 @@
 RMDIR=rmdir
 TAR=tar
 CHMOD=chmod
-GZIP=gzip -9
+# the GZIP variable is reserved by gzip program
+GZIP_BIN=gzip -9
 GUNZIP=gunzip
 ZIP=zip
 UNIX2DOS=unix2dos
@@ -387,8 +388,8 @@
 install_man::
 	-mkdir -p $(DESTDIR)$(man1dir)
 	-mkdir -p $(DESTDIR)$(man5dir)
-	for i in $(MAN_PAGE_1_LIST); do cat $(srcdir)/doc/man/$$i.1 | $(SED) 's/@version@/$(VERSION)/g' | $(GZIP) -c >  $(DESTDIR)$(man1dir)/$$i.1.gz; chmod 644 $(DESTDIR)$(man1dir)/$$i.1.gz; done
-	for i in $(MAN_PAGE_5_LIST); do cat $(srcdir)/doc/man/$$i.5 | $(GZIP) -c >  $(DESTDIR)$(man5dir)/$$i.5.gz; chmod 644 $(DESTDIR)$(man5dir)/$$i.5.gz; done
+	for i in $(MAN_PAGE_1_LIST); do cat $(srcdir)/doc/man/$$i.1 | $(SED) 's/@version@/$(VERSION)/g' | $(GZIP_BIN) -c >  $(DESTDIR)$(man1dir)/$$i.1.gz; chmod 644 $(DESTDIR)$(man1dir)/$$i.1.gz; done
+	for i in $(MAN_PAGE_5_LIST); do cat $(srcdir)/doc/man/$$i.5 | $(GZIP_BIN) -c >  $(DESTDIR)$(man5dir)/$$i.5.gz; chmod 644 $(DESTDIR)$(man5dir)/$$i.5.gz; done
 
 download_dlx: $(DLXLINUX_TAR)
 
@@ -407,7 +408,7 @@
 	$(RM) -rf $(DESTDIR)$(sharedir)/dlxlinux
 	cp -r dlxlinux $(DESTDIR)$(sharedir)/dlxlinux
 	$(CHMOD) 755 $(DESTDIR)$(sharedir)/dlxlinux
-	$(GZIP) $(DESTDIR)$(sharedir)/dlxlinux/hd10meg.img
+	$(GZIP_BIN) $(DESTDIR)$(sharedir)/dlxlinux/hd10meg.img
 	$(CHMOD) 644 $(DESTDIR)$(sharedir)/dlxlinux/*
 	for i in bochs-dlx; do cp $(srcdir)/build/linux/$$i $(bindir)/$$i; $(CHMOD) 755 $(DESTDIR)$(bindir)/$$i; done
 
@@ -441,7 +442,7 @@
 	$(SHELL) ./build/win32/cpp2cc
 
 tar:
-	NAME=`pwd|$(SED) 's/.*\///'`; (cd ..; $(RM) -f $$NAME.zip; tar cf - $$NAME | $(GZIP) > $$NAME.tar.gz); ls -l ../$$NAME.tar.gz
+	NAME=`pwd|$(SED) 's/.*\///'`; (cd ..; $(RM) -f $$NAME.zip; tar cf - $$NAME | $(GZIP_BIN) > $$NAME.tar.gz); ls -l ../$$NAME.tar.gz
 
 zip:
 	NAME=`pwd|$(SED) 's/.*\///'`; (cd ..; $(RM) -f $$NAME.zip; $(ZIP) $$NAME.zip -r $$NAME -x \*CVS\* -x \*.cvsignore ); ls -l ../$$NAME.zip
