diff -ur bochs.old/config.h.in bochs/config.h.in
--- bochs.old/config.h.in	2004-08-28 10:41:46.000000000 +0200
+++ bochs/config.h.in	2004-09-03 10:23:22.000000000 +0200
@@ -903,7 +903,7 @@
 #ifdef BX_USE_ETH_ARPBACK
 #  define ETH_ARPBACK 1
 #endif
-#if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) || defined(__APPLE__)
+#if defined(__FreeBSD__) || defined(__FreeBSD_kernel__) || defined(__OpenBSD__) || defined(__NetBSD__) || defined(__NetBSD_kernel__) || defined(__APPLE__)
 #define ETH_FBSD  1
 #endif
 #if defined(linux)
diff -ur bochs.old/iodev/cdrom.cc bochs/iodev/cdrom.cc
--- bochs.old/iodev/cdrom.cc	2004-08-30 12:47:09.000000000 +0200
+++ bochs/iodev/cdrom.cc	2004-09-03 10:23:22.000000000 +0200
@@ -85,7 +85,7 @@
 #include "cdrom_beos.h"
 #define BX_CD_FRAMESIZE 2048
 
-#elif (defined (__NetBSD__) || defined(__OpenBSD__) || defined(__FreeBSD__))
+#elif (defined(__NetBSD__) || defined(__NetBSD_kernel__) || defined(__OpenBSD__) || defined(__FreeBSD__) || defined(__FreeBSD_kernel__))
 // OpenBSD pre version 2.7 may require extern "C" { } structure around
 // all the includes, because the i386 sys/disklabel.h contains code which 
 // c++ considers invalid.
@@ -721,7 +721,7 @@
   // Spin up the cdrom drive.
 
   if (fd >= 0) {
-#if defined(__NetBSD__)
+#if defined(__NetBSD__) || defined(__NetBSD_kernel__)
     if (ioctl (fd, CDIOCSTART) < 0)
        BX_DEBUG(( "start_cdrom: start returns error: %s", strerror (errno) ));
     return(true);
@@ -740,7 +740,7 @@
   // some ioctl() calls to really eject the CD as well.
 
   if (fd >= 0) {
-#if (defined(__OpenBSD__) || defined(__FreeBSD__))
+#if (defined(__OpenBSD__) || defined(__FreeBSD__) || defined(__FreeBSD_kernel__))
     (void) ioctl (fd, CDIOCALLOW);
     if (ioctl (fd, CDIOCEJECT) < 0)
 	  BX_DEBUG(( "eject_cdrom: eject returns error." ));
@@ -962,7 +962,7 @@
 
   return true;
   }
-#elif (defined(__NetBSD__) || defined(__OpenBSD__) || defined(__FreeBSD__))
+#elif (defined(__NetBSD__) || defined(__NetBSD_kernel__) || defined(__OpenBSD__) || defined(__FreeBSD__) || defined(__FreeBSD_kernel__))
   {
   struct ioc_toc_header h;
   struct ioc_read_toc_entry t;
@@ -1161,7 +1161,7 @@
   
     return(buf.st_size);
   }
-#elif (defined(__NetBSD__) || defined(__OpenBSD__))
+#elif (defined(__NetBSD__) || defined(__NetBSD_kernel__) || defined(__OpenBSD__))
   {
   // We just read the disklabel, imagine that...
   struct disklabel lp;
@@ -1227,7 +1227,7 @@
   return(num_sectors);
 
   }
-#elif defined(__FreeBSD__)
+#elif defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
   {
   // Read the TOC to get the size of the data track.
   // Keith Jones <freebsd.dev@blueyonder.co.uk>, 16 January 2000
diff -ur bochs.old/iodev/eth_tap.cc bochs/iodev/eth_tap.cc
--- bochs.old/iodev/eth_tap.cc	2004-08-06 17:49:54.000000000 +0200
+++ bochs/iodev/eth_tap.cc	2004-09-03 10:23:22.000000000 +0200
@@ -100,9 +100,9 @@
 #include <sys/socket.h>
 #include <sys/uio.h>
 #include <sys/wait.h>
-#if defined(__FreeBSD__) || defined(__APPLE__)  // Should be fixed for other *BSD
+#if defined(__FreeBSD__) || defined(__FreeBSD_kernel__) || defined(__APPLE__)  // Should be fixed for other *BSD
 #include <net/if.h>
-#else
+#elif defined(__linux__)
 #include <asm/types.h>
 #include <linux/netlink.h>
 #include <linux/if.h>
@@ -269,7 +269,7 @@
   Bit8u txbuf[BX_PACKET_BUFSIZ];
   txbuf[0] = 0;
   txbuf[1] = 0;
-#if defined(__FreeBSD__) || defined(__APPLE__)  // Should be fixed for other *BSD
+#if defined(__FreeBSD__) || defined(__FreeBSD_kernel__) || defined(__APPLE__)  // Should be fixed for other *BSD
   memcpy (txbuf, buf, io_len);
   unsigned int size = write (fd, txbuf, io_len);
   if (size != io_len) {
@@ -318,7 +318,7 @@
   nbytes = read (fd, buf, sizeof(buf));
 
   // hack: discard first two bytes
-#if defined(__FreeBSD__) || defined(__APPLE__)  // Should be fixed for other *BSD
+#if defined(__FreeBSD__) || defined(__FreeBSD_kernel__) || defined(__APPLE__)  // Should be fixed for other *BSD
   rxbuf = buf;
 #else
   rxbuf = buf+2;
diff -ur bochs.old/iodev/serial.h bochs/iodev/serial.h
--- bochs.old/iodev/serial.h	2004-07-28 21:36:42.000000000 +0200
+++ bochs/iodev/serial.h	2004-09-03 10:25:08.000000000 +0200
@@ -40,7 +40,7 @@
 #include "serial_raw.h"
 #endif // USE_RAW_SERIAL
 
-#if defined(__NetBSD__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__linux__) || defined(__GNU__) || defined(__APPLE__)
+#if defined(__NetBSD__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__linux__) || defined(__GNU__) || defined(__GLIBC__) || defined(__APPLE__)
 #define SERIAL_ENABLE
 extern "C" {
 #include <termios.h>
diff -ur bochs.old/iodev/soundlnx.cc bochs/iodev/soundlnx.cc
--- bochs.old/iodev/soundlnx.cc	2004-06-19 17:20:14.000000000 +0200
+++ bochs/iodev/soundlnx.cc	2004-09-03 10:25:48.000000000 +0200
@@ -28,7 +28,7 @@
 
 
 #include "iodev.h"
-#if (defined(linux) || defined(__FreeBSD__)) && BX_SUPPORT_SB16
+#if (defined(linux) || defined(__FreeBSD__) || defined(__FreeBSD_kernel__)) && BX_SUPPORT_SB16
 #define LOG_THIS bx_sb16.
 
 #include <errno.h>
diff -ur bochs.old/iodev/soundlnx.h bochs/iodev/soundlnx.h
--- bochs.old/iodev/soundlnx.h	2002-12-24 11:12:26.000000000 +0100
+++ bochs/iodev/soundlnx.h	2004-09-03 10:23:22.000000000 +0200
@@ -27,7 +27,7 @@
 // This file (SOUNDLNX.H) written and donated by Josef Drexler
 
 
-#if (defined(linux) || defined(__FreeBSD__))
+#if (defined(linux) || defined(__FreeBSD__) || defined(__FreeBSD_kernel__))
 
 #include "bochs.h"
 
diff -ur bochs.old/plex86/plex86.h bochs/plex86/plex86.h
--- bochs.old/plex86/plex86.h	2003-01-10 05:27:51.000000000 +0100
+++ bochs/plex86/plex86.h	2004-09-03 10:23:22.000000000 +0200
@@ -156,7 +156,7 @@
  *  ioctl() names.
  */
 
-#if defined(__linux__) || defined(__NetBSD__) || defined(__FreeBSD__)
+#if defined(__linux__) || defined(__NetBSD__) || defined(__NetBSD_kernel__) || defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 #ifdef __linux__
 #include <asm/ioctl.h>
 #else
