diff -ur bochs.old/.bochsrc bochs/.bochsrc
--- bochs.old/.bochsrc	2004-09-01 20:12:21.000000000 +0200
+++ bochs/.bochsrc	2004-09-03 10:54:11.000000000 +0200
@@ -17,7 +17,7 @@
 # the "wx" display library.
 #=======================================================================
 #config_interface: textconfig
-#config_interface: wx
+config_interface: wx
 
 #=======================================================================
 # DISPLAY_LIBRARY
@@ -31,15 +31,9 @@
 #
 # The choices are: 
 #   x              use X windows interface, cross platform
-#   win32          use native win32 libraries
-#   carbon         use Carbon library (for MacOS X)
-#   beos           use native BeOS libraries
-#   macintosh      use MacOS pre-10
-#   amigaos        use native AmigaOS libraries
 #   sdl            use SDL library, cross platform
 #   svga           use SVGALIB library for Linux, allows graphics without X11
 #   term           text only, uses curses/ncurses library, cross platform
-#   rfb            provides an interface to AT&T's VNC viewer, cross platform
 #   wx             use wxWindows library, cross platform
 #   nogui          no display at all
 #
@@ -50,16 +44,10 @@
 # Some display libraries now support specific option to control their
 # behaviour. See the examples below for currently supported options.
 #=======================================================================
-#display_library: amigaos
-#display_library: beos
-#display_library: carbon
-#display_library: macintosh
 #display_library: nogui
-#display_library: rfb, options="timeout=60" # time to wait for client
 #display_library: sdl, options="fullscreen" # startup in fullscreen mode
 #display_library: term
-#display_library: win32, options="legacyF12" # use F12 to toggle mouse
-#display_library: wx
+display_library: wx
 #display_library: x
 
 #=======================================================================
@@ -70,10 +58,10 @@
 # You can also use the environment variable $BXSHARE to specify the
 # location of the BIOS.
 #=======================================================================
-romimage: file=$BXSHARE/BIOS-bochs-latest, address=0xf0000
-#romimage: file=bios/BIOS-bochs-2-processors, address=0xf0000
-#romimage: file=bios/BIOS-bochs-4-processors, address=0xf0000
-#romimage: file=bios/rombios.bin, address=0xf0000
+romimage: file=/usr/share/bochs/BIOS-bochs-latest, address=0xf0000
+#romimage: file=/usr/share/bochs/BIOS-bochs-2-processors, address=0xf0000
+#romimage: file=/usr/share/bochs/BIOS-bochs-4-processors, address=0xf0000
+#romimage: file=/usr/share/bochs/BIOS-bochs-8-processors, address=0xf0000
 
 #=======================================================================
 # MEGS
@@ -109,9 +97,8 @@
 # VGAROMIMAGE
 # You now need to load a VGA ROM BIOS into C0000.
 #=======================================================================
-#vgaromimage: file=bios/VGABIOS-lgpl-latest
-#vgaromimage: file=bios/VGABIOS-elpin-2.40
-vgaromimage: file=$BXSHARE/VGABIOS-elpin-2.40
+vgaromimage: /usr/share/vgabios/vgabios.bin
+#vgaromimage: /usr/share/vgabios/vgabios.debug.bin
 
 #=======================================================================
 # FLOPPYA:
@@ -218,11 +205,8 @@
 #   ata3-master: type=disk, mode=flat, path=483M.sample, cylinders=1024, heads=15, spt=63
 #   ata3-slave:  type=cdrom, path=iso.sample, status=inserted
 #=======================================================================
-ata0-master: type=disk, mode=flat, path="30M.sample", cylinders=615, heads=6, spt=17
-#ata0-slave: type=cdrom, path=D:, status=inserted
-#ata0-slave: type=cdrom, path=/dev/cdrom, status=inserted
-#ata0-slave: type=cdrom, path="drive", status=inserted
-#ata0-slave: type=cdrom, path=/dev/rcd0d, status=inserted 
+#ata0-master: type=disk, path="$HOME/.bochs/guest.hd0", cylinders=#cylinders#\, heads=16, spt=63
+ata0-slave: type=cdrom, path=/dev/#cdrom#, status=inserted
 
 #=======================================================================
 # BOOT:
@@ -237,7 +221,8 @@
 #   boot: a
 #=======================================================================
 #boot: floppy
-boot: disk
+#boot: disk
+boot: cdrom
 
 #=======================================================================
 # IPS:
@@ -314,8 +299,9 @@
 #   log: ./bochs.out
 #   log: /dev/tty
 #=======================================================================
+log: /dev/stdout
 #log: /dev/null
-log: bochsout.txt
+#log: /tmp/bochsout.txt
 
 #=======================================================================
 # LOGPREFIX:
@@ -392,7 +378,7 @@
 #   com2: enabled=1, mode=file, dev=serial.out
 #   com3: enabled=1, mode=raw, dev=com1
 #=======================================================================
-#com1: enabled=1, mode=term, dev=/dev/ttyp9
+com1: enabled=1, mode=term, dev=/dev/#com#
 
 
 #=======================================================================
@@ -408,7 +394,7 @@
 #   parport2: enabled=1, file="/dev/lp0"
 #   parport1: enabled=0
 #=======================================================================
-parport1: enabled=1, file="parport.out"
+parport1: enabled=1, file="/dev/lp0"
 
 #=======================================================================
 # SB16:
@@ -445,7 +431,7 @@
 # For an example look at the next line:
 #=======================================================================
 
-#sb16: midimode=1, midi=/dev/midi00, wavemode=1, wave=/dev/dsp, loglevel=2, log=sb16.log, dmatimer=600000
+sb16: midimode=1, midi=/dev/midi00, wavemode=1, wave=/dev/dsp, loglevel=2, log=/dev/stdout, dmatimer=600000
 
 #=======================================================================
 # VGA_UPDATE_INTERVAL:
@@ -530,18 +516,6 @@
 private_colormap: enabled=0
 
 #=======================================================================
-# fullscreen: ONLY IMPLEMENTED ON AMIGA
-#             Request that Bochs occupy the entire screen instead of a 
-#             window.
-#
-# Examples:
-#   fullscreen: enabled=0
-#   fullscreen: enabled=1
-#=======================================================================
-#fullscreen: enabled=0
-#screenmode: name="sample"
-
-#=======================================================================
 # ne2k: NE2000 compatible ethernet adapter
 #
 # Examples:
@@ -569,10 +543,7 @@
 # Bochs execution. The network interface name is supplied to the script
 # as first parameter
 #=======================================================================
-# ne2k: ioaddr=0x240, irq=9, mac=fe:fd:00:00:00:01, ethmod=fbsd, ethdev=en0 #macosx
-# ne2k: ioaddr=0x240, irq=9, mac=b0:c4:20:00:00:00, ethmod=fbsd, ethdev=xl0
-# ne2k: ioaddr=0x240, irq=9, mac=b0:c4:20:00:00:00, ethmod=linux, ethdev=eth0
-# ne2k: ioaddr=0x240, irq=9, mac=b0:c4:20:00:00:01, ethmod=win32, ethdev=MYCARD
+# ne2k: ioaddr=0x240, irq=9, mac=b0:c4:20:00:00:00, ethmod=#kernel#, ethdev=#eth#
 # ne2k: ioaddr=0x240, irq=9, mac=fe:fd:00:00:00:01, ethmod=tap, ethdev=tap0
 # ne2k: ioaddr=0x240, irq=9, mac=fe:fd:00:00:00:01, ethmod=tuntap, ethdev=/dev/net/tun0, script=./tunconfig
 
@@ -585,7 +556,7 @@
 # Examples:
 #   keyboard_mapping: enabled=1, map=gui/keymaps/x11-pc-de.map
 #=======================================================================
-keyboard_mapping: enabled=0, map=
+#keyboard_mapping: enabled=0, map=/usr/share/bochs/keymaps/x11-pc-de.map
 
 #=======================================================================
 # KEYBOARD_TYPE:
@@ -638,114 +609,3 @@
 #-------------------------
 #pcidev: vendor=0x1234, device=0x5678
 
-#=======================================================================
-# for Macintosh, use the style of pathnames in the following
-# examples.
-#
-# vgaromimage: :bios:VGABIOS-elpin-2.40
-# romimage: file=:bios:BIOS-bochs-latest, address=0xf0000
-# floppya: 1_44=[fd:], status=inserted
-#=======================================================================
-
-
-#=======================================================================
-#
-# The following directives are DEPRECATED 
-# Please convert them to the new syntax  or remove them
-#
-#=======================================================================
-
-#=======================================================================
-#
-# The DISKC option is deprecated. Use ATA* options instead.
-#
-# DISKC: file=, cyl=, heads=, spt=
-# Point this at a hard disk image file.  To create 
-# a hard disk image, try running bximage.  It will help you choose the
-# size and then suggest a diskc line that works with it.
-#
-# In UNIX it may be possible to use a raw device as a Bochs hard disk, 
-# but WE DON'T RECOMMEND IT.  In Windows there is no easy way.
-#
-# Examples:
-#   diskc: file=10M.sample, cyl=306, heads=4, spt=17
-#   diskc: file=20M.sample, cyl=615, heads=4, spt=17
-#   diskc: file=30M.sample, cyl=615, heads=6, spt=17
-#   diskc: file=46M.sample, cyl=940, heads=6, spt=17
-#   diskc: file=62M.sample, cyl=940, heads=8, spt=17
-#   diskc: file=112M.sample, cyl=900, heads=15, spt=17
-#   diskc: file=483M.sample, cyl=1024, heads=15, spt=63
-#=======================================================================
-#diskc: file="30M.sample", cyl=615, heads=6, spt=17
-
-#=======================================================================
-#
-# The DISKD option is deprecated. Use ATA* options instead.
-#
-# DISKD:
-# See DISKC above for syntax
-#
-# NOTE: diskd and cdromd must not be used together!
-#=======================================================================
-#diskd: file="diskd.img", cyl=615, heads=6, spt=17
-
-#=======================================================================
-#
-# The CDROMD option is deprecated. Use ATA* options instead.
-#
-# CDROMD:
-#
-# cdromd: dev=/dev/cdrom, status=inserted
-# cdromd: dev=/dev/cdrom, status=ejected
-# cdromd: dev=e:, status=ejected
-#
-# In windows, the drive letter + colon notation should be used for cdroms.
-# Depending on versions of windows and drivers, you may only be able to 
-# access the "first" cdrom in the system.  On MacOSX, use path="drive"
-# to access the physical drive.
-#
-# NOTE: diskd and cdromd must not be used together!
-#=======================================================================
-#cdromd: dev=D:, status=inserted
-#cdromd: dev=/dev/cdrom, status=inserted
-#cdromd: dev="drive", status=inserted
-
-#=======================================================================
-#
-# The TIME0 directive is DEPRECATED. Use the CLOCK directive instead
-#
-# TIME0:
-# Specifies the start (boot) time of the virtual machine. Use a time 
-# value as returned by the time(2) system call. If no time0 value is 
-# set or if time0 equal to 1 (special case), the simulation will be
-# started at the current time of the host. 
-#
-# Examples:
-#   time0: 1           # Now
-#   time0: 315529200   # Tue Jan  1 00:00:00 1980
-#   time0: 631148400   # Mon Jan  1 00:00:00 1990
-#   time0: 938581955   # Wed Sep 29 07:12:35 1999
-#   time0: 946681200   # Sat Jan  1 00:00:00 2000
-#=======================================================================
-#time0: 938581955
-
-#=======================================================================
-#
-# The PIT directive is DEPRECATED. Use the CLOCK directive instead
-#
-# PIT:
-# The PIT is the programmable interval timer.  It has an option that tries to
-# keep the PIT in sync with real time.  This feature is still experimental,
-# but it may be useful if you want to prevent Bochs from running too fast, for
-# example a DOS video game.  Be aware that with the realtime pit option, your
-# simulation will not be repeatable; this can a problem if you are debugging.
-#=======================================================================
-#pit: realtime=1
-
-#=======================================================================
-# NEWHARDDRIVESUPPORT: enabled=[0|1]
-# The old harddrive code is not maintened any more.
-# Default value is enabled=1
-#=======================================================================
-#newharddrivesupport: enabled=1
-
Only in bochs/: .bochsrc.orig
Only in bochs/: .bochsrc.rej~
Only in bochs/: .bochsrc~
diff -ur bochs.old/configure.in bochs/configure.in
--- bochs.old/configure.in	2004-08-18 13:48:56.000000000 +0200
+++ bochs/configure.in	2004-09-03 10:48:53.000000000 +0200
@@ -1646,6 +1646,29 @@
     ]
   )
 
+AC_MSG_CHECKING(for 3DNOW support)
+AC_ARG_ENABLE(3dnow,
+  [  --enable-3dnow                      3DNOW support],
+  [if test "$enableval" = yes; then
+    AC_MSG_RESULT(yes)
+    AC_DEFINE(BX_SUPPORT_3DNOW, 1)
+   elif test "$enableval" = no; then
+    AC_MSG_RESULT(no)
+    AC_DEFINE(BX_SUPPORT_3DNOW, 0)
+   fi
+   ],
+  [
+    if test "$use_x86_64" = 1; then
+      AC_MSG_RESULT(yes)
+      AC_DEFINE(BX_SUPPORT_3DNOW, 1)
+    else
+      AC_MSG_RESULT(no)
+      AC_DEFINE(BX_SUPPORT_3DNOW, 0)
+    fi
+    AC_MSG_RESULT(no)
+    ]
+  )
+
 AC_MSG_CHECKING(for SEP support)
 AC_ARG_ENABLE(sep,
   [  --enable-sep                      SYSENTER/SYSEXIT support],
diff -ur bochs.old/doc/man/bochs.1 bochs/doc/man/bochs.1
--- bochs.old/doc/man/bochs.1	2002-12-13 20:46:53.000000000 +0100
+++ bochs/doc/man/bochs.1	2004-09-03 10:48:53.000000000 +0200
@@ -57,8 +57,8 @@
 .SH LICENSE
 This program  is distributed  under the terms of the  GNU
 Lesser General Public License as published  by  the  Free
-Software  Foundation.  See  the  COPYING file located  in
-/usr/local/share/doc/bochs/ for details on the license and
+Software  Foundation.  See  the LGPL file located  in
+/usr/share/common-licenses/ for details on the license and
 the lack of warrantee.
 .\"SKIP_SECTION"
 .SH AVAILABILITY
diff -ur bochs.old/misc/sb16/sb16ctrl.c bochs/misc/sb16/sb16ctrl.c
--- bochs.old/misc/sb16/sb16ctrl.c	2001-10-03 15:10:38.000000000 +0200
+++ bochs/misc/sb16/sb16ctrl.c	2004-09-03 10:48:53.000000000 +0200
@@ -43,7 +43,7 @@
 #	include <dos.h>
 #endif
 
-#ifdef linux
+#ifdef __unix__
 #	include <sys/io.h>
 #	include <errno.h>
 #	define inp inb
@@ -80,7 +80,7 @@
 /* Enable access to the emulator port */
 void enableport()
 {
-#ifdef linux
+#ifdef __unix__
   if (ioperm(EMULPORT, 1, 1)) {
     printf("Could not access emulator port %03x: %s.\n", EMULPORT, strerror(errno));
     exit(1);
